src/main/java/apoc/util/s3util/S3ParamsExtractor.java

```diff
@@ -0,0 +1,71 @@
+package apoc.util.s3util;
+
+import com.amazonaws.regions.Regions;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Objects;
+
+public class S3ParamsExtractor {
+    public static S3Params extract(URL url) throws IOException, IllegalArgumentException {
+
+        if (!"s3".equals(url.getProtocol())) {
+            throw new IllegalArgumentException("Unsupported protocol '" + url.getProtocol() + "'");
+        }
+
+        //aws credentials
+        String accessKey = null;
+        String secretKey = null;
+
+        if (url.getUserInfo() != null) {
+            String[] credentials = url.getUserInfo().split("[:]");
+            accessKey = credentials[0];
+            secretKey = credentials[1];
+        } else {
+            if (url.getQuery() != null) {
+                String[] queryStrings = url.getQuery().split("&");
+                if (queryStrings.length == 2) {
+                    for (String query : queryStrings) {
+                        String[] parts = query.split("=");
+                        if (parts.length == 2) {
+                            String valueDecoded = parts[1];
+                            if ("accessKey".equals(parts[0])) {
+                                accessKey = valueDecoded;
+                            }
+                            if ("secretKey".equals(parts[0])) {
+                                secretKey = valueDecoded;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // bucket
+        String bucket = url.getHost().substring(0, url.getHost().indexOf("."));
+
+        // key
+        String key = url.getPath().substring(1);
+
+        // endpoint
+        String endpoint = url.getHost().substring(url.getHost().indexOf(".") + 1);
+
+        String region = null;
+
+        for (Regions r: Regions.values()){
+            if(endpoint.contains(r.getName())){
```

Comment:
what about case, we should probably use tolower on both. And probably split the endpoint on "." ? and access the right segment.